### DISCUSSION TITLE
State of the Ally

### DISCUSSION THREAD

**Discussion by seerge** [seerge/g-helper#1985] [Announcements]
Hello fellow community,

In the last few updates I have added a lot of new features for the Ally handheld :

- Controller modes (Auto /  Gamepad / Mouse)
- Custom controller bindings
- Deadzone and vibration settings 
- Improved backlight control, including super low brightness level
- FPS limiter
- Brightness control via DPad Left / Right
- Realtime battery discharge readings
- And many others :)

What are the most important features you are still missing from Armoury ? **Let me know.**

Looking ahead, there are few things that I intentionally **don't plan to add** (cause they will make app too heavy or very difficult to do)

- Game Launcher (cause this is a category of it's own, and there are a lot of standalone and very complicated game launchers on the market already like Playnite)
- Per Game profiles (is not really doable without a game launcher itself)

![image](https://github.com/seerge/g-helper/assets/5920850/1f6c36de-8f5d-48bd-aad7-7ef5c9d21bdf)

**Comment by leal-expelled7** [comment]
Really appreciate the work lately to further improve support and features for the ROG Ally, @seerge!

The big feature currently missing in G-Helper for me is the ability to manually toggle the controller modes via the Command Center OEM button, for example. The best use case for this is in games that have controller support but where a quick switch to Mouse mode would let you accomplish something much quicker.

**Reply by seerge** [reply]
@k-trahan hello,

This build should allow binding any controller key to ``Controller Mode`` toggle. Keep in mind that toggle with turn off "Auto" mode and will just switch between Mouse / Gamepad

[GHelper.zip](https://github.com/seerge/g-helper/files/14077405/GHelper.zip)
![Screenshot 2024-01-28 201041](https://github.com/seerge/g-helper/assets/5920850/c5e6dd94-56c1-4129-bd3b-8c1a4083f8b1)

**Reply by leal-expelled7** [reply]
Thank you, this works very well but is only bindable through the Controller menu. Would it be possible to bind it to the "Ctrl Center" button either via the "Extra" menu or by adding the Ally OEM buttons to the "Controller" menu?

One other potential bug I notice is that the Reset button on the Controller menu doesn't seem to reset any controller binds nor does it disable the "Disable Controller" option. It DOES reset the deadzone sliders so that may be working as intended.

Finally, another issue I notice is that the Min/33% LED brightness option no longer works on 0.150.0. I had to revert back to 0.135.0 to get it working again. Went back to 0.150.0/the above build again and 33% doesn't work. This is with Asus services both completely disabled and enabled. If you prefer I don't include bug reports here, no problem, I can file via Issues instead.

Thank you again!

**Reply by leal-expelled7** [reply]
Update to above, my apologies but 33% IS working, I just had too dark of a color selected for it to not show as off at 33%. However, the Rainbow pattern does not work on any brightness level for me. Static, Breathe, Color Cycle, and Strobe are working but Rainbow does not.

**Reply by seerge** [reply]
@k-trahan this build should have controller mode toggle also under Extra, even tho I plan to move even ROG/CtrlCenter bindings completely to "Controller" screen in future anyway

[GHelper.zip](https://github.com/seerge/g-helper/files/14077615/GHelper.zip)

As for Resetting controller, it doens't touch bindings atm :) but it's intentional

**Reply by leal-expelled7** [reply]
Love it, thank you so much!

**Reply by seerge** [reply]
I would advice to bind it as a secondary action (i.e. with M-key) to Right stick click. Then it's intuitive toggle between mouse mode for that stick and gamepad.

**Comment by Kenzim** [comment]
Thank you so much for your work on supporting the ROG ally.

One thing I miss it having gyro support for the controller, I suppose this would involve some virtual controller to implement? Not sure if that would be outside of the scope for something like this. I changed to handheld companion from armoury crate, but lately feeling like even handheld companion is too clunky/bloated for the few things I really need, like tdp control. I can't get the m1/m2 to map with it either, which ghelper can now do without issue. The only thing I really lose from HC is the gyro, but I'm hoping perhaps rewasd will add support for this.

**Reply by seerge** [reply]
@Kenzim well, gyro could be very difficult :) 

Currently G-Helper does all bindings / settings for controller on native level using same commands as Armoury sends to it. So it doesn't "emulate" anything, it just sets and forgets.

For gyro, I can imagine, you will have to listen to the sensors in realtime and somehow emit appropriate controller commands all the time

**Reply by seerge** [reply]
Update, I have checked how AC does it. And it is as expected - it reads gyro/accelerometer data in real-time, and just spams commands to controller to emit actual joystick moves. 

Emitting moves is actually easy, but I have no idea how to read accelerometer data. And I also don't know hot to read joystick keypresses neither (and it's needed if you want to turn on gyro-action only on right-trigger for example)

**Reply by jassunca** [reply]
@seerge I assume that's because all you can do is just intercept raw data from the accelerometer to AC, and with so many variable data coming from it, you have no idea what is what to translate it into joystick movement?
I've been having a look through [Bosch's sensor api for the BMI323 IMU](https://github.com/boschsensortec/BMI323_SensorAPI) (which apparently is the IMU sensor the Ally uses for gyro/accelerometer), trying to figure exactly what type of values it works with, but with my very limited knowledge in C, I haven't had much luck - even after I did manage to compile and run it, I don't know which interface to use to connect to it, seems that by default it expects a data input from a separate application board controller, rather than the integrated I2C one. It could be a way to try and know what calls and data returns to look out for, but I guess that until someone figures how it communicates in the first place, we're stumped.

**Comment by lawrstay** [comment]
Huge thanks for the support on Ally community @seerge , sending some support your way to return the love.  🤟 

I'd love to be able to bind the "**Toggle XG Mobile**" to a custom keyboard shortcut instead of the Ally's controller. 
My Ally is mostly docked to my TV, and control via an external gamepad. If I can bind the "Toggle XGM" to a keyboard hotkey, I can use REWASD to map that hotkey to any external gamepad button comob (eg: L1 + L2 + R1 + R2) 

I've been using the toggle and it works pretty well being mapped to M2, therefore this request is more of an additive to my previous request. Many thanks again, hope I'm not being too demanding.

**Reply by seerge** [reply]
@lawrstay you actually can do that already. 

Just press ``Ctrl + Shift + Alt + F3``, this is what controller emits as well, when you set that binding. And app handles this hotkey as a command for XGM :)

**Reply by lawrstay** [reply]
This is amazing, and works perfectly. Thank you seerge!

**Comment by Logidoggy** [comment]
GHelper v0.151 is flagging as a trojan in Windows Defender. I'm not sure if it's a false flag but it's continually quarantined each time I attempt to download it wither through the .zip file or as a standalone. Just thought I'd give a heads up in case your github was hacked.

**Reply by seerge** [reply]
@Logidoggy hello, please [check FAQ](https://github.com/seerge/g-helper/wiki/FAQ#windows-defender-marks-app-as-malware--virus)

> ### Windows Defender marks app as malware / virus
> False positives from Windows Defender (or any other similar system that uses machine learning for detection) is possible as the application is not digitally signed with a certificate. You can always download a version below or compile the app by yourself. All application sources are open and can be monitored from A to Z :)

**Comment by noFEARgr** [comment]
Hello and thank you for the wonderful support of the rog ally.
The latest command center has an option to change on the fly the tdp , do you think it is possible to have the same option for g-helper ?
Once again thanks !

**Reply by seerge** [reply]
@noFEARgr hello, but you can drag TDP sliders in G-Helper from day one ? they apply in realtime

**Comment by LadyHylia** [comment]
Thank you so much for making this, I've been looking for something to replace armory crate since day one! I do have two {small?} suggestions I hope could be implemented; move the rog and control center button mapping to the controller section with the rest of the buttons, and add the xbox guide as a default option to map them to. This is all I want out of this program but I can't figure out how to do it as is.

**Reply by seerge** [reply]
@LadyHylia hello :) I plan to move ROG/ControlCenter buttons bindings to Controller screen somewhere soon.

As for binding them to Guide / XBOX button - it doesn't seem possible on this stage, as they are not part of that hardware xbox controller that Ally imitates (you also can't bind them in AC for same reason) :(

**Reply by LadyHylia** [reply]
That's strange, because you can bind the guide button to the other buttons in AC. Is it because the rog and control center buttons aren't part of xinput/whatever the rest of the device uses? If that's the case, could something like an autohotkey script be used to capture a semi-random input then change it to the guide button?

**Reply by seerge** [reply]
@LadyHylia you can bind it to any other button in G-Helper as well. This is because whole "binding" process happens on hardware level. I.e. you tell controller to emit certain thing when certain button is pressed and then it will be like that until reboot. 

G-Helper does exactly same as AC in this sense.

ROG / ControlCenter buttons are different, they work in a same way as macro buttons (M1-M5) work on laptops. That's the reason they are in Extra settings in a first place.

**Reply by LadyHylia** [reply]
Darn, well that's a shame :( I guess the only other thing I could think of is maybe to have it so that when these buttons are "bound" to the guide button(or any other xinput button I guess), it would swap the bind of one of the system's other buttons(like A or something) to the guide, tell the system to send an input for _that_ button(assuming that's even possible), and then quickly revert its' bind to whatever it was before. That would be really janky though haha

**Reply by seerge** [reply]
I'm just binding ``M + Menu`` or ``M + Select`` button to XBox :) it's convenient as you don't use XBox button that often anyway

**Reply by LadyHylia** [reply]
I like to have my m buttons bound to face buttons usually, so that I don't need to move my thumb off the right stick to, say, jump. Is there a way to bind the m buttons to their own binding, and have them activate the rest of the map's secondary functions simultaneously?

**Reply by seerge** [reply]
@LadyHylia no, as soon as you bind M directly to something, it stops working as a secondary modifier

**Reply by LadyHylia** [reply]
That's what I thought :( I guess I've been spoiled by the steam controller with its ability to map basically anything to anything. I'll have to get used to using one m button for button inputs and one to input shift. Thank you for your time

**Reply by dxcotoxb** [reply]
To use the ac button as xbox button. I assigned it win+g. Its the same... i think

**Comment by lawrstay** [comment]
Hello seerge, do you think it would be a good idea to have separate "Memory Assigned to GPU" settings for iGPU vs eGPU ?
Some games runs much better on 5GB with iGPU on Ally, however when connected the XGM, it could benefit from additional memory from the eGPU, but that requires manual changing of the Memory whenever switching from iGPU to eGPU. I'm hoping that by having separate settings, it allows hotswapping memory assignment without needing a system restart for a more seamless experience. Thank you for considering this suggestion.

**Reply by seerge** [reply]
@lawrstay hello, adjusting Memory Assigned to GPU requires reboot anyway, so there is no way to do it on the fly anyhow.

**Comment by crabnebula007** [comment]
The possible Ally Controller modes are Auto/Gamepad/Mouse/Skip. What does Skip do?

**Reply by seerge** [reply]
@ericlitvak pretty much what it says - G-helper would just skip controller settings all together. This is needed only if you want to set up controller in some other way (for example using AC itself or so)

**Reply by crabnebula007** [reply]
Many thanks for your reply, that's what I thought but I wasn't sure and for now I'm still using AC for the gyro support. Thanks for all your work!

**Reply by sirkuttin** [reply]
if we don't have AC running and have this set to skip, is this the same as Off in AC? Sometimes i want teh controller off so my BT gamepad is more easily registered as controller 1

**Reply by seerge** [reply]
@sirkuttin no, it’s not the same.

But you can completely turn off controller from “Controller” settings screen with appropriate checkbox

**Comment by skeud51** [comment]
Hello, thank you for the application. I play on the rog ally x. For me the only reason not to use your application.there is no gyroscope included from armory creat.I often play FPS games and it's essential.
Could you in the future add the possibility of activating the gyroscope?

**Reply by seerge** [reply]
@skeud51 hello,

The main problem is that it seem to be no open source info on how to read gyro-sensor data.

I know how to emit joystick movements tho, so if that would be possible - it would be possible to program it.

However there is a workaround - a reWASD app that can handle gyro :) so you can try it in conjunction with G-Helper

**Comment by magnifylifeacademy-collab** [comment]
Hi I just wanted to ask, are you planning to update g helper to bring controller support to the Xbox ally x soon? I saw there’s no controller section when using it on the new Xbox ally x.
